include/mesh/FoamDataStore.h

```diff
@@ -0,0 +1,346 @@
+#pragma once
+
+#include <DataIO.h>
+
+#include <fvCFD_moose.h>
+
+// This method extracts the keys associated with fields of type T from the
+// mesh object registry. Note for some fields, the field.name() and the
+// key are not the same. *strict* indicates whether types derived from T are
+// collected
+template <typename T, bool strict>
+inline std::vector<Foam::string>
+getFieldkeys(const Foam::fvMesh & mesh)
+{
+  std::vector<Foam::string> fieldKeyList;
+  for (auto key : mesh.toc())
+  {
+    if (mesh.foundObject<T>(key))
+    {
+      auto & field = mesh.lookupObjectRef<T>(key);
+      if (!field.isOldTime() &&
+          ((Foam::isType<T>(field) && strict) || (Foam::isA<T>(field) && !strict)))
+      {
+        fieldKeyList.push_back(key);
+      }
+    }
+  }
+
+  return fieldKeyList;
+}
+
+template <class Type, class GeomMesh>
+inline void
+readBoundary(istream & stream, Foam::DimensionedField<Type, GeomMesh> & field)
+{
+  (void)stream;
+  (void)field;
+}
+
+template <typename GeoField>
+inline void
+readBoundary(istream & stream, GeoField & field)
+{
+  for (auto & bField : field.boundaryFieldRef())
+  {
+    std::vector<typename GeoField::value_type> data(bField.size());
+    loadHelper(stream, data, nullptr);
+    for (auto i = 0lu; i < data.size(); ++i)
+      bField[i] = data[i];
+  }
+}
+
+// readField for GeometricFields and DimensionedFields
+template <typename GeoField>
+inline GeoField &
+readField(std::istream & stream,
+          Foam::fvMesh & mesh,
+          GeoField * parent_field,
+          unsigned int old_time_index)
+{
+  assert(old_time_index == 0 || parent_field);
+
+  std::string field_name;
+  loadHelper(stream, field_name, nullptr);
+  auto & field = (old_time_index > 0 && parent_field) ? parent_field->oldTimeRef(old_time_index)
+                                                      : mesh.lookupObjectRef<GeoField>(field_name);
```

Comment:
It might be cleaner to have two different `readField`s, one accepting `GeoField & parent_field` and one `old_time_index`